### PR TITLE
ci(.github): refactor build workflow to add sca, signing and provenance for images and artifacts

### PR DESCRIPTION
**Pull request by saisatishkarra**
### Checklist prior to review

- [ ] [Link to relevant issue][1] as well as docs and UI issues --
- [ ] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [ ] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [ ] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [ ] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

### Overview
- Refactors build pipeline to integrate
   -  Reusable docker images TAR / artifact binary files across workflow jobs using upload / download v4 action
   - Scanning (SCA) of docker images
   - Signing of docker image manifests recursively for all platforms and publishing signatures to docker hub (registry needs to be created)
   - Build provenance for artifacts and images being produced using slsa-generators
      - Provenance is NOT generated on PR events and hence cannot test it until merged 

### TODO
- Test the workflow for first on kumahq and sync changes to kong-mesh '
- Update any kong-mesh / child repo workflow sync files for the refactored piepline
- Create `kumahq/notary` public docker repository to push image signatures for verification.

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

**Comment by saisatishkarra** [comment]
@jijiechen @lahabana Is there a way to safely test this POC refactored workflow? I would need review/fixes/approval before proceeding to update the sync files for EE.

**Comment by jijiechen** [comment]
> Is there a way to safely test this POC refactored workflow? I would need review/fixes/approval before proceeding to update the sync files for EE.
I see you've turned on running action on your personal repo. That's what I was doing when I was changing workflow files. Please note that the secrets are not available for PR builds, to test steps involving secrets, we should use branch/tag builds in our own repo. For thrid party connections, create personal connections if required and use them to replace the real ones when testing.

**Comment by saisatishkarra** [comment]
> > Is there a way to safely test this POC refactored workflow? I would need review/fixes/approval before proceeding to update the sync files for EE.
> 
> I see you've turned on running action on your personal repo. That's what I was doing when I was changing workflow files. Please note that the secrets are not available for PR builds, to test steps involving secrets, we should use branch/tag builds in our own repo. For third party connections, create personal connections if required and use them to replace the real ones when testing.

@jijiechen The actions are disabled in my forked repo and to test this I would need to mock many things like secrets / docker repo / update docker org/pulp server which is not feasible. Why can this not be triggered under kumahq/kuma repo with a `ci/force-publish` label from my branch for testing and push preview tags as usual to pulp/docker repos?

File: .github/workflows/build-test-distribute.yaml [hunk 1]
```diff
@@ -199,6 +388,72 @@ jobs:
...
+  # runs-on option is [not supported](https://github.com/orgs/community/discussions/25783)
+  # ENV option is [not supported](https://github.com/orgs/community/discussions/26671)
+  # Reusable workflow doesn't support exrernal COSIGN_REPOSITORY via input / env variable
+  # TODO:
+  #   Split provenance jobs for internal / official releases when repositories are split
+  images-provenance:
+    name: Images Provenance
+    needs: [metadata, distribute-images, sign-image-manifests, provenance-metadata]
+    if: >
+      github.repository_owner == 'Kong' && github.event_name != 'pull_request' && needs.metadata.outputs.ALLOW_PUSH == 'true'
```
Comment (lahabana) [inline .github/workflows/build-test-distribute.yaml:427]:
```suggestion
      needs.metadata.outputs.ALLOW_PUSH == 'true'
``` 

This should be good enough no?

File: mk/docker.mk [hunk 2]
```diff
@@ -18,6 +18,14 @@ KUMA_IMAGES = $(call build_image,$(IMAGES_RELEASE) $(IMAGES_TEST))
 images/show: ## output all images that are built with the current configuration
 	@echo $(KUMA_IMAGES)
 
+.PHONY: manifests/release/show
+manifests/release/show:
+	@echo $(call build_image,$(IMAGES_RELEASE))
+
+.PHONY: images/release/show
+images/release/show:
+	@echo $(foreach arch,$(ENABLED_GOARCHES), $(patsubst %, %-$(arch), $(call build_image,$(IMAGES_RELEASE))))
```
Comment (lahabana) [inline mk/docker.mk:27]: To check later: verify what happens with UBI images and this

File: .github/workflows/build-test-distribute.yaml [hunk 3]
```diff
@@ -140,14 +152,88 @@ jobs:
...
+      - name: Scan Image
+        id: image_sbom
+        uses: Kong/public-shared-actions/security-actions/scan-docker-image@4119b418bba4462593b5238718772356f84f67ee
+        with:
+          asset_prefix: ${{ steps.img_meta.outputs.tar_name }}
+          image: ${{ matrix.tar }}
+  distribute-images:
+    needs: ["metadata", "scan-images"]
+    if: >
+      github.repository_owner == 'Kong'  && needs.metadata.outputs.ALLOW_PUSH == 'true'
```
Comment (lahabana) [inline .github/workflows/build-test-distribute.yaml:199]:
```suggestion
     needs.metadata.outputs.ALLOW_PUSH == 'true'
```

That's enough no?

File: .github/workflows/build-test-distribute.yaml [hunk 4]
```diff
@@ -71,34 +71,55 @@ jobs:
           make dev/tools
       - run: |
           make test
-  distributions:
-    timeout-minutes: 40
-    needs: ["check", "test", "test_e2e", "test_e2e_env"]
-    if: ${{ always() }}
+  metadata:
```
Comment (lahabana) [inline .github/workflows/build-test-distribute.yaml:74]: NIT: if you have outputs in `build` as well I'd collapse these 2 actions (it makes the graph a little simpler to follow)

File: .github/workflows/build-test-distribute.yaml [hunk 5]
```diff
@@ -71,34 +71,55 @@ jobs:
...
         run: |
           echo '${ENABLED_GOARCHES}|${ENABLED_GOOSES}' > .run-full-matrix
       - name: "Maybe set flag to push build artifacts"
+        id: artifacts_metadata
         if: github.event_name == 'push' || contains(github.event.pull_request.labels.*.name, 'ci/force-publish')
         run: |
-          echo 'ALLOW_PUSH=true' >> $GITHUB_ENV
+          echo 'ALLOW_PUSH=true' >> $GITHUB_OUTPUT
+  build:
+    needs: ["metadata", "test", "test_e2e", "test_e2e_env"]
```
Comment (lahabana) [inline .github/workflows/build-test-distribute.yaml:101]: I'd have this only depend on `check` so it'd make the build generally quicker as build would be in parallel

File: .github/workflows/build-test-distribute.yaml [hunk 6]
```diff
@@ -140,14 +152,88 @@ jobs:
...
+    steps:
+      - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+        with:
+          fetch-depth: 0
+      - uses: actions/setup-go@0c52d547c9bc32b1aa3301fd7a9cb496313a4491 # v5.0.0
+        with:
+          go-version-file: go.mod
+          cache-dependency-path: |
+            .run-full-matrix
+            go.sum
```
Comment (lahabana) [inline .github/workflows/build-test-distribute.yaml:217]: Don't think you need go in this action now

File: .github/workflows/build-test-distribute.yaml [hunk 7]
```diff
@@ -160,6 +246,109 @@ jobs:
...
+  # TODO:
+  #   Split sigining step and notary registries to avoid customers viewing signatures of internal images
+  #     Official image signatures: kumahq/notary
+  #     Internal/Preview image signatures:kumahq/notary-internal
+  sign-image-manifests:
+    permissions:
+      id-token: write # needed for signing the images with GitHub OIDC Token
+    needs: ["metadata", "distribute-images"]
+    if: >
+      github.repository_owner == 'Kong'  && needs.metadata.outputs.ALLOW_PUSH == 'true'
```
Comment (lahabana) [inline .github/workflows/build-test-distribute.yaml:269]:
```suggestion
      needs.metadata.outputs.ALLOW_PUSH == 'true'
```

File: .github/workflows/build-test-distribute.yaml [hunk 8]
```diff
@@ -160,6 +246,109 @@ jobs:
...
+
+    runs-on: ubuntu-latest-kong
```
Comment (lahabana) [inline .github/workflows/build-test-distribute.yaml:271]: not going to work for OSS (we have stuff to make this run on our runners for KM).
```suggestion
    runs-on: ubuntu-latest
```

File: .github/workflows/build-test-distribute.yaml [hunk 9]
```diff
@@ -160,6 +246,109 @@ jobs:
...
+          make docker/login
+      - name: sign image manifest
+        id: attest
+        uses: Kong/public-shared-actions/security-actions/sign-docker-image@4119b418bba4462593b5238718772356f84f67ee
+        with:
+          image_digest: ${{ steps.manifest_meta.outputs.digest }}
+          tags: ${{ matrix.image_manifest }}
+          signature_registry: ${{ env.NOTARY_REPOSITORY }}
+        env:
+          NOTARY_REPOSITORY: kumahq/notary # TODO
```
Comment (lahabana) [inline .github/workflows/build-test-distribute.yaml:296]: Ideally extract this at a high level to make it easier to override in KM

File: mk/docker.mk [hunk 10]
```diff
@@ -92,7 +100,7 @@ docker/$(1)/$(2)/push:
 endef
 $(foreach goarch, $(SUPPORTED_GOARCHES),$(foreach image, $(IMAGES_RELEASE) $(IMAGES_TEST),$(eval $(call DOCKER_TARGETS_BY_ARCH,$(image),$(goarch)))))
 
-# create and push a manifest for each
+# create manifest and push a manifest for each component
```
Comment (lahabana) [inline mk/docker.mk:103]:
```suggestion
# create and push a manifest for each component
```

File: .github/workflows/build-test-distribute.yaml [hunk 11]
```diff
@@ -199,6 +388,72 @@ jobs:
...
+  # The provenance job is reusable workflow that:
+  # Cannot be used as a step within other jobs
+  # Cannot be used to pass ENV variables
+  # GHA limitations don't allow dynamic output keys / expressions
+  provenance-metadata:
+    name: Provenance metadata
+    needs: [metadata, distributions, distribute-images, sign-image-manifests]
+    runs-on: ubuntu-latest-kong
+    if: >
+      github.repository_owner == 'Kong'  && needs.metadata.outputs.ALLOW_PUSH == 'true' && github.event_name != 'pull_request'
```
Comment (lahabana) [inline .github/workflows/build-test-distribute.yaml:400]:
```suggestion
      needs.metadata.outputs.ALLOW_PUSH == 'true'
```

**Review by lahabana — COMMENTED** [review]
Thanks for starting on this. Is there any change we could most of these in their own workflow and use `workflow_call` or something like this? `build-test-distribute` is already pretty hard to follow...

File: .github/workflows/build-test-distribute.yaml (hunk 1 above)
Comment (saisatishkarra) [reply .github/workflows/build-test-distribute.yaml:427]: `needs.metadata.outputs.ALLOW_PUSH == 'true' and github.event_name != 'pull_request'` is needed due to the [limitation](https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/container#supported-triggers) mentione here for supported triggers

Comment (saisatishkarra) [reply .github/workflows/build-test-distribute.yaml:427]: Does `needs.metadata.outputs.ALLOW_PUSH == 'true'` guarantee to run this job only for `kumahq` owned repos?

File: mk/distribution.mk [hunk 12]
```diff
@@ -99,6 +100,10 @@ ENABLED_DIST_NAMES=$(filter $(addprefix %,$(ENABLED_ARCH_OS)),$(foreach elt,$(DI
 .PHONY: build/distributions ## Build tar.gz for each enabled distribution
 build/distributions: $(patsubst %,build/distributions/out/$(DISTRIBUTION_TARGET_NAME)-%.tar.gz,$(ENABLED_DIST_NAMES))
 
+.PHONY: build/distribution-provenance-metadata
+build/distributions-provenance-metadata:
+	cd build/distributions/out; sha256sum *.tar.gz | base64 -w0 > artifact_digest_file.text
```
Comment (lahabana) [inline mk/distribution.mk:105]: Maybe
```suggestion
	cd build/distributions/out; sha256sum *.tar.gz | base64 -w0 > $(BUILD_DIR)/artifact_digest_file.text
```

File: .github/workflows/build-test-distribute.yaml (hunk 1 above)
Comment (saisatishkarra) [reply .github/workflows/build-test-distribute.yaml:427]: done

File: .github/workflows/build-test-distribute.yaml (hunk 3 above)
Comment (saisatishkarra) [reply .github/workflows/build-test-distribute.yaml:199]: done

File: .github/workflows/build-test-distribute.yaml (hunk 4 above)
Comment (saisatishkarra) [reply .github/workflows/build-test-distribute.yaml:74]: done. removed `metadata` job

File: .github/workflows/build-test-distribute.yaml (hunk 5 above)
Comment (saisatishkarra) [reply .github/workflows/build-test-distribute.yaml:101]: fixed to depend on `["check"]` to make build job in parallel with tests

File: .github/workflows/build-test-distribute.yaml (hunk 7 above)
Comment (saisatishkarra) [reply .github/workflows/build-test-distribute.yaml:269]: done

File: .github/workflows/build-test-distribute.yaml (hunk 8 above)
Comment (saisatishkarra) [reply .github/workflows/build-test-distribute.yaml:271]: done

File: .github/workflows/build-test-distribute.yaml (hunk 9 above)
Comment (saisatishkarra) [reply .github/workflows/build-test-distribute.yaml:296]: done. Point this to "{{github.repository_owner}}/notary" since docker hub org is the same as GitHub org to avoid overriding for KM

File: mk/docker.mk (hunk 10 above)
Comment (saisatishkarra) [reply mk/docker.mk:103]: done

File: .github/workflows/build-test-distribute.yaml (hunk 11 above)
Comment (saisatishkarra) [reply .github/workflows/build-test-distribute.yaml:400]: done

File: mk/distribution.mk (hunk 12 above)
Comment (saisatishkarra) [reply mk/distribution.mk:105]: updated.

**Comment by saisatishkarra** [comment]
closing this PR in favor of #9169 for testing purposes. Any reviews must be posted on #9169
